output/kafka/client.go

```diff
@@ -0,0 +1,436 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package kafka
+
+import (
+	"errors"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Shopify/sarama"
+
+	"github.com/eapache/go-resiliency/breaker"
+
+	"github.com/elastic/beats/v7/libbeat/common/fmtstr"
+	"github.com/elastic/beats/v7/libbeat/outputs/codec"
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/elastic-agent-shipper-client/pkg/proto/messages"
+)
+
+type Client struct {
+	log           *logp.Logger
+	hosts         []string
+	topic         *fmtstr.EventFormatString
+	key           *fmtstr.EventFormatString
+	index         string
+	codec         codec.Codec
+	config        sarama.Config
+	mux           sync.Mutex
+	done          chan struct{}
+	recordHeaders []sarama.RecordHeader
+
+	asyncProducer sarama.AsyncProducer
+	wg            sync.WaitGroup
+
+	//observer outputs.Observer             TODO: what to do with observers?
```

Comment:
We will need to get metrics for outputs in the shipper. I'll defer to @faec and @leehinman on whether we should try to port over the Beat observer pattern or do something else.

Reply:
I think we need a design doc, made an issue for it.  https://github.com/elastic/elastic-agent-shipper/issues/163